File: 309. Best Time to Buy and Sell Stock with Cooldown/Solution.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    Map<String, Integer> profits = new HashMap<>();

    public int maxProfit(int[] prices) {
        return dfs(prices, 0, true);
    }

    public int dfs(int[] prices, int index, boolean buying) {
        if (index >= prices.length)
            return 0;
        else if (profits.containsKey(index + "" + buying))
            return profits.get(index + "" + buying);
        else {
            int coolDown = dfs(prices, index + 1, buying);
            int val;
            if (buying) {
                val = dfs(prices, index + 1, !buying) - prices[index];
            } else {
                val = dfs(prices, index + 2, !buying) + prices[index];
            }
            profits.put(index + "" + buying, Math.max(val, coolDown));
        }
        return profits.get(index + "" + buying);
    }
}

File: 309. Best Time to Buy and Sell Stock with Cooldown/Solution.py
from typing import List


class Solution:

    def maxProfit(self, prices: List[int]) -> int:
        return self.dfs(prices, 0, True, {})

    def dfs(self, prices: List[int], index: int, buying: bool, profits) -> int:
        if (index >= len(prices)):
            return 0
        if (index, buying) in profits:
            return profits[(index, buying)]
        coolDown = self.dfs(prices, index+1, buying, profits)
        if (buying):
            result = self.dfs(prices, index+1, not buying,
                              profits)-prices[index]
        else:
            result = self.dfs(prices, index+2, not buying,
                              profits)+prices[index]
        profits[(index, buying)] = max(result, coolDown)
        return profits[(index, buying)]
